Whisky/Views/Bottle/Pins/PinCreationView.swift

```diff
@@ -0,0 +1,124 @@
+    @State private var newPinName: String = ""
+    @State private var newPinURL: URL = BottleData.defaultBottleDir
+    @State private var isValid: Bool = false
+    @State private var isDuplicate: Bool = false
+    @Environment(\.dismiss) private var dismiss
+
+    var body: some View {
+        VStack(alignment: .leading) {
+            Text("pin.title")
+                .bold()
+            Divider()
+            HStack {
+                Text("pin.name")
+                    .frame(maxWidth: /*@START_MENU_TOKEN@*/.infinity/*@END_MENU_TOKEN@*/, alignment: .leading)
+                TextField(String(), text: $newPinName)
+                    .frame(width: 180)
+                    .textFieldStyle(.roundedBorder)
+            }
+            HStack {
+                Text("pin.path")
+                    .frame(maxWidth: /*@START_MENU_TOKEN@*/.infinity/*@END_MENU_TOKEN@*/, alignment: .leading)
+                Button("create.browse") {
+                    let panel = NSOpenPanel()
+                    panel.canChooseFiles = true
+                    panel.allowedContentTypes = [UTType.exe,
+                                                 UTType(exportedAs: "com.microsoft.msi-installer"),
+                                                 UTType(exportedAs: "com.microsoft.bat")]
+                    panel.directoryURL = newPinURL
+                    panel.canChooseDirectories = false
+                    panel.allowsMultipleSelection = false
+                    panel.canCreateDirectories = false
+                    panel.begin { result in
+                        if result == .OK {
+                            if let url = panel.urls.first {
+                                let oldUrl = newPinURL.lastPathComponent
+                                                      .replacingOccurrences(of: ".exe", with: "")
+                                                      .replacingOccurrences(of: ".msi", with: "")
+                                                      .replacingOccurrences(of: ".bat", with: "")
+                                newPinURL = url
+                                // Only reset newPinName if the textbox hasn't been modified
+                                if newPinName.isEmpty || newPinName == oldUrl {
+                                    newPinName = url.lastPathComponent
+                                                    .replacingOccurrences(of: ".exe", with: "")
+                                                    .replacingOccurrences(of: ".msi", with: "")
+                                                    .replacingOccurrences(of: ".bat", with: "")
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            let pinPath = newPinURL.prettyPath()
+            Text(pinPath)
+                .truncationMode(.middle)
+                .lineLimit(2, reservesSpace: true)
+                .help(pinPath)
```

Comment:
just a preference but I think you can make newPinURL optional and not display the path if nothing is selected. It's a bit confusing to see a path for something you haven't selected yet.

I'll leave it to your discretion as I'm not a UX designer:

You will have to fix a few places:
1. You use an alert with `presenting` to deal with the optional url. You can leave the actions blank for the default "ok"
2. remove the `onAppear`
3. Deal with optional in panel results. Putting `?` should do the trick
4. Coalesce (`??`) a default panel directory in case nothing is yet selected: (i.e. `panel.directoryURL = newPinURL ?? bottle.url.appending(path: "drive_c")`

Reply:
I was waffling on this anyway, so I took this as the impetus to just do it.